### ISSUE TITLE
DllImport of MacOS frameworks not working on Big Sur

### ISSUE THREAD

**Issue by jherby2k** [dotnet/runtime#62337]
### Description

I guess with Big Sur onwards, Apple has hidden framework assemblies. You're supposed to be able to dlopen them as normal, however DllImport does not appear to work. I've tried p/invoking DlOpen directly (not sure if DllImport attempts to check for the file's presence before calling DlOpen - that would cause problems) but even P/Invoking DlOpen fails to return a handle. This worked under Catalina et al, where the libraries are visible. Any suggestions or is this a known bug? Couldn't find anything online.

I've tried with .NET 6 arm64 and x86 versions of the SDK and MacOS itself.

### Reproduction Steps

```c#
        [DllImport("/System/Library/Frameworks/CoreAudio.framework/CoreAudio", CallingConvention = CallingConvention.Cdecl, ExactSpelling = true)]
        internal static extern AudioFileStatus AudioFileInitializeWithCallbacks(
            IntPtr userData,
            NativeCallbacks.AudioFileReadCallback readCallback,
            NativeCallbacks.AudioFileWriteCallback writeCallback,
            NativeCallbacks.AudioFileGetSizeCallback getSizeCallback,
            NativeCallbacks.AudioFileSetSizeCallback setSizeCallback,
            AudioFileType fileType,
            ref AudioStreamBasicDescription description,
            uint flags,
            out AudioFileHandle handle);
```

### Expected behavior

the imported function works.

### Actual behavior

fails to load the library

### Other information

Typical Apple, i can't find any official documentation of this change. But there are online discussions - https://mjtsai.com/blog/2020/06/26/reverse-engineering-macos-11-0/

basically it sounds like dlopen should internally load the libraries from the dyld_shared_cache. It should be possible to manually extract the libraries using a tool and then redistributing it with my app, but that's not ideal - rather just use the one on the system.

**Comment by jkoritzinsky** [comment]
cc: @AaronRobinsonMSFT @elinor-fung @lambdageek

**Comment by lambdageek** [comment]
@jherby2k Hi, do you have a full example that we could try?

I tried the following (.net 6.0 on x64 Big Sur 11.6.1) in a `dotnet new console` template:

```csharp
using System;
using System.Runtime.InteropServices;


public class  Imp {
    [DllImport("/System/Library/Frameworks/CoreAudio.framework/CoreAudio", CallingConvention = CallingConvention.Cdecl, ExactSpelling = true)]
    public static extern /*AudioFileStatus*/ int AudioFileInitializeWithCallbacks(
	IntPtr userData,
	/*NativeCallbacks.AudioFileReadCallback*/ IntPtr readCallback,
	/*NativeCallbacks.AudioFileWriteCallback*/ IntPtr writeCallback,
	/*NativeCallbacks.AudioFileGetSizeCallback*/ IntPtr getSizeCallback,
	/*NativeCallbacks.AudioFileSetSizeCallback*/ IntPtr setSizeCallback,
	/*AudioFileType*/int fileType,
	/*ref AudioStreamBasicDescription*/ IntPtr description,
	uint flags,
	/*out AudioFileHandle*/ out IntPtr handle);
}

public class P {
    public static void Main ()
    {
	Console.WriteLine ("before loading");
	IntPtr handle = IntPtr.Zero;
	Imp.AudioFileInitializeWithCallbacks(userData: IntPtr.Zero, readCallback: IntPtr.Zero, writeCallback: IntPtr.Zero, getSizeCallback: IntPtr.Zero, setSizeCallback: IntPtr.Zero, fileType: 0, description: IntPtr.Zero, flags: 0, handle: out handle);
	Console.WriteLine ("after loading");
    }
}
```

Than ran with ```DYLD_PRINT_LIBRARIES=1 dotnet run```

The output is:

```
...
dyld: loaded: <1BEF817F-F64C-366E-AA4B-103FB346370A> /usr/local/share/dotnet/shared/Microsoft.NETCore.App/6.0.0/libclrjit.dylib
dyld: loaded: <EACA4343-FA6E-3041-AD5F-3E51260DE0DB> /usr/local/share/dotnet/shared/Microsoft.NETCore.App/6.0.0/libSystem.Native.dylib
before loading
dyld: loaded: <7178815E-1890-3A3D-B0FD-F93EBB529E78> /System/Library/Frameworks/CoreAudio.framework/Versions/A/CoreAudio
dyld: loaded: <523B9C03-A39E-3FF4-8216-34309C158113> /System/Library/PrivateFrameworks/AppServerSupport.framework/Versions/A/AppServerSupport
dyld: loaded: <B80E66C3-5A5F-391A-9475-15DB16BA72E8> /System/Library/PrivateFrameworks/perfdata.framework/Versions/A/perfdata
dyld: loaded: <968E021C-1BCD-3249-BBAD-FEE64370EB1F> /System/Library/PrivateFrameworks/AssertionServices.framework/Versions/A/AssertionServices
dyld: loaded: <92B2E9B6-7FA8-3BE5-AD1A-25A2BC36821E> /System/Library/PrivateFrameworks/AudioToolboxCore.framework/Versions/A/AudioToolboxCore
...
after loading
```

So as far as the dynamic linker is concerned, we loaded the library.

(I also checked that it works with .NET 6 Mono, by adding `<SelfContained>true</SelfContained<UseMonoRuntime>true</UseMonoRuntime><RuntimeIdentifier>osx-x64</RuntimeIdentifier>` to the .csproj, and runnign with `MONO_LOG_LEVEL=debug MONO_LOG_MASK=dll dotnet run`)

**Comment by jherby2k** [comment]
Thanks, yeah, it seems I had a different problem with Big Sur, misdiagnosed when I couldn't find the libs.
